### PR TITLE
Extend db clients with drop partitions capabilities

### PR DESCRIPTION
**Pull request by stoyanr**
Signed-off-by: Plamen Doychev <plamen.doychev@sap.com>

### CODE REVIEW

File: etc/concourse/scripts/cf-deploy-stage [hunk 1]
```diff
@@ -34,4 +34,5 @@ if [ "$BIND_DB_SERVICE" == "true" ]; then
   bindService ${ABACUS_PREFIX}abacus-usage-reporting $DB_AGGREGATOR,$DB_ACCUMULATOR
   bindService ${ABACUS_PREFIX}abacus-cf-bridge,${ABACUS_PREFIX}abacus-cf-renewer $DB_BRIDGE
   bindService ${ABACUS_PREFIX}abacus-provisioning-plugin,${ABACUS_PREFIX}abacus-account-plugin $DB_PLUGINS
+  bindService ${ABACUS_PREFIX}abacus-housekeeper $DB_COLLECTOR,$DB_METER,$DB_ACCUMULATOR,$DB_AGGREGATOR,$DB_BRIDGE,$DB_PLUGINS
```
Comment (hsiliev) [inline etc/concourse/scripts/cf-deploy-stage]: Hey shouldn't this be a part of another PR that introduces the housekeeper app?

File: lib/utils/couchclient/src/index.js [hunk 2]
```diff
@@ -699,65 +738,84 @@ const couchclient = (part, uri, cons) => {
...
+  };
+
+  if(/:/.test(server))
+    dropServerDatabases(server, regex, {}, localOnly, lcb);
+  else
+    dropInMemoryDatabases(server, regex, inMemoryOpt, lcb);
+};
+
+// Delete all partitions that match the specified regex
+const deletePartitions = (server, regex, cb) => {
```
Comment (hsiliev) [inline lib/utils/couchclient/src/index.js]: Not sure if partition has any meaning in Couch. Perhaps we can leave only `drop` for couch?

File: lib/utils/couchclient/src/index.js [hunk 3]
```diff
@@ -699,65 +738,84 @@ const couchclient = (part, uri, cons) => {
...
+  dropDatabases(server, regex, false, cb);
```
Comment (hsiliev) [inline lib/utils/couchclient/src/index.js]: Running demo or smoke test vs productive instance will result in dropped database. We should disable this code-line by default and enable it only on request (via env.variable for instance) to protect productive data from accidental deletion.

File: lib/utils/couchclient/src/index.js [hunk 4]
```diff
@@ -107,6 +107,45 @@ const key = (uri) => {
   return undefined;
 };
 
+// List all databases on the specified server
```
Comment (stoyanr) [inline lib/utils/couchclient/src/index.js]: It's there because this is the order in the mongoclient. And in the mongoclient, the order is such because these methods are also used elsewhere (e.g. findServer).

File: etc/concourse/scripts/cf-deploy-stage (hunk 1 above)
Comment (stoyanr) [inline etc/concourse/scripts/cf-deploy-stage]: OK, I will move that to a different PR.

File: lib/utils/couchclient/src/index.js (hunk 2 above)
Comment (stoyanr) [inline lib/utils/couchclient/src/index.js]: "Partition" is a unit that's used for partitioning. With Couch, partitions are always databases. With Mongo, however, partitions can be databases or collections, and `deletePartitions` will correctly determine what to delete in both cases. Therefore, the naming "partition" is important and since both clients should have the same API, both should have such a method.

The method `deletePartitions` is intentionally different from `drop` because:
* It always works with partitions, which might be different from databases (with Mongo)
* It deletes partitions also on remote servers, while `drop` deletes databases on local servers only.

File: lib/utils/couchclient/src/index.js (hunk 3 above)
Comment (stoyanr) [inline lib/utils/couchclient/src/index.js]: No. This method is not used (and should not be used) anywhere except the new `housekeeper` module which will be part of `cf-abacus-broker`.

I will add more comments in the code to warn about the implications of using this method. In the future, we can also add an `eslint` check to make it even harder to misuse it accidentally.
